pkg/networkpath/traceroute/tcp/utils.go

```diff
@@ -0,0 +1,430 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tcp
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"go.uber.org/multierr"
+	"golang.org/x/net/ipv4"
+)
+
+const (
+	// URG is the urgent TCP flag
+	URG = 1 << 5
+	// ACK is the acknowledge TCP flag
+	ACK = 1 << 4
+	// PSH is the push TCP flag
+	PSH = 1 << 3
+	// RST is the reset TCP flag
+	RST = 1 << 2
+	// SYN is the synchronization TCP flag
+	SYN = 1 << 1
+	// FIN is the final TCP flag
+	FIN = 1 << 0
+)
+
+type (
+	// canceledError is sent when a listener
+	// is canceled
+	canceledError string
+
+	// icmpResponse encapsulates the data from
+	// an ICMP response packet needed for matching
+	icmpResponse struct {
+		SrcIP        net.IP
+		DstIP        net.IP
+		TypeCode     layers.ICMPv4TypeCode
+		InnerSrcIP   net.IP
+		InnerDstIP   net.IP
+		InnerSrcPort uint16
+		InnerDstPort uint16
+		InnerSeqNum  uint32
+	}
+
+	// tcpResponse encapsulates the data from a
+	// TCP response needed for matching
+	tcpResponse struct {
+		SrcIP       net.IP
+		DstIP       net.IP
+		TCPResponse *layers.TCP
+	}
+
+	rawConnWrapper interface {
+		SetReadDeadline(t time.Time) error
+		ReadFrom(b []byte) (*ipv4.Header, []byte, *ipv4.ControlMessage, error)
+		WriteTo(h *ipv4.Header, p []byte, cm *ipv4.ControlMessage) error
+	}
+)
+
+func localAddrForHost(destIP net.IP, destPort uint16) (*net.UDPAddr, error) {
+	// this is a quick way to get the local address for connecting to the host
+	// using UDP as the network type to avoid actually creating a connection to
+	// the host, just get the OS to give us a local IP and local ephemeral port
+	conn, err := net.Dial("udp4", net.JoinHostPort(destIP.String(), strconv.Itoa(int(destPort))))
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	localAddr := conn.LocalAddr()
+
+	localUDPAddr, ok := localAddr.(*net.UDPAddr)
+	if !ok {
+		return nil, fmt.Errorf("invalid address type for %s: want %T, got %T", localAddr, localUDPAddr, localAddr)
+	}
+
+	return localUDPAddr, nil
+}
+
+// createRawTCPPacket creates a TCP packet with the specified parameters
+func createRawTCPPacket(sourceIP net.IP, sourcePort uint16, destIP net.IP, destPort uint16, seqNum uint32, ttl int, flags byte) (*ipv4.Header, []byte, error) {
+	ipHdr := ipv4.Header{
+		Version:  4,
+		Len:      20,
+		TTL:      ttl,
+		ID:       418218,
+		Protocol: 6, // TCP
+		Dst:      destIP,
+		Src:      sourceIP,
+	}
+
+	// Create TCP packet with the specified flags
+	// and sequence number
+	tcpPacket := make([]byte, 20)
```

Comment:
Could you use something like this https://pkg.go.dev/github.com/google/gopacket@v1.1.19/layers#TCP for constructing a tcp packet?